src/bin_pattern.jl

```diff
@@ -0,0 +1,169 @@
+ ⋅  2  2
+ ⋅  ⋅  3
+```
+"""
+function bin_sparse_matrix!(M::AbstractArray{T}, M_id::AbstractSparseMatrixCSC, max_num_bins::Integer) where {T}
+    @assert max_num_bins >= 0 "max_num_bins must be non-negative"
+    @assert size(M, 1) == size(M_id, 1) && size(M, 2) == size(M_id, 2) "M must be the same size as M_id"
+
+    if T <: Integer
+        M = 1.0*M
+    end
+    P = eltype(M)
+
+    m, n = size(M_id)
+    M_i, M_j = findnz(M_id)
+    M_pat_nnz = count(!iszero, M_id)
+
+    separated_at = zero(P)
+
+    if max_num_bins == 0
+        data = Vector(1:M_pat_nnz)
+        M_id[:, :] = sparse(M_i, M_j, data, m, n)
+    else
+        nz_M = zeros(M_pat_nnz)
+
+        for k in 1:M_pat_nnz
+            nz_M[k] = M[M_i[k], M_j[k]]
+        end
+        min_left, max_left, min_right, max_right = separated_min_max(nz_M, separated_at)
+        bin_ids = binned_with_separated_min_max(nz_M, max_num_bins, min_left, max_left, min_right, max_right, separated_at)
+        new_ids = bin_mapping(bin_ids)
+        M_id[:, :] = sparse(M_i, M_j, new_ids, m, n)
+    end
+    return M_id
+end
+
+_extrema(v) = isempty(v) ? (Inf, -Inf) : extrema(v)
+
+function separated_min_max(v::AbstractVector{T}, separated_at::T) where {T<:Real} 
+    min_left, max_left = _extrema(v[v .<= separated_at])
+    min_right, max_right = _extrema(v[v .>= separated_at])
+
+    return min_left, max_left, min_right, max_right
+end
+
+
+function binned_with_separated_min_max(v::AbstractVector{T}, max_num_bins::Integer,
+    min_left::Real, max_left::Real, min_right::Real, max_right::Real, separated_at::Real) where{T}
+    n = length(v)
+5
+    if min_left <= max_left
+        @assert all(min_left .<= v) "some element of v is less than min_left"
+        @assert min_left <= separated_at "incompatible min_left and separated_at"
+    end
+    if min_right <= max_right
+        @assert all(v .<= max_right) "some element of v is greater than max_right"
+        @assert separated_at <= max_right "incompatible separated_at and max_right"
+    end
+
+    if max_left == min_right
+        loc_max_num_left_right_bins = max_num_bins - 1
+    else
+        loc_max_num_left_right_bins = max_num_bins
+    end
+
+    if separated_at > min_left
+        left_dist = separated_at - min_left
+    else
+        left_dist = 0
+    end
+
+    if max_right > separated_at
+        right_dist = max_right - separated_at
+    else
+        right_dist = 0
+    end
+
+    total_dist = left_dist + right_dist
+
+    if total_dist != 0
+        max_n_left_bins = floor(Int, (loc_max_num_left_right_bins * left_dist) / total_dist)
+        max_n_right_bins = floor(Int, (loc_max_num_left_right_bins * right_dist) / total_dist)
+    else
+        max_n_left_bins = Inf
+        max_n_right_bins = Inf
+    end
+    @assert max_n_left_bins + max_n_right_bins <= loc_max_num_left_right_bins
+
+    inv_h_l = max_n_left_bins / left_dist
+    inv_h_r = max_n_right_bins / right_dist
+
+    fuzz = 1E2  # magic constant from matlab
```

Comment:
If I'm not mistaken, `1E2` will be a `Float64`. Making this of Type `T` (?) might make the code much faster.